### PR TITLE
#903 Avoid building the timeline DB when not requested in the job

### PR DESCRIPTION
**Pull request by e2l-su**
This PR to resolve issue 903

### CODE REVIEW

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp [hunk 1]
```diff
@@ -972,16 +974,18 @@ ModelLine::evalZ(const double& t) {
                 throw DYNError(Error::MODELER, DynamicLineStatusNotSupported, id_);
               switch (getConnectionState()) {
               case OPEN:
-                network_->addEvent(id_, DYNTimeline(LineCloseSide2));
+                if (network_->hasTimeline()) network_->addEvent(id_, DYNTimeline(LineCloseSide2));
                 modelBus2_->getVoltageLevel()->connectNode(modelBus2_->getBusIndex());
                 break;
               case CLOSED:
-                network_->addEvent(id_, DYNTimeline(LineOpenSide1));
+                if (network_->hasTimeline()) network_->addEvent(id_, DYNTimeline(LineOpenSide1));
```
Comment (lecourtoisflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: It is more readable to put this in 2 lines

File: dynawo/sources/Modeler/Common/DYNSubModel.h [hunk 2]
```diff
@@ -730,6 +730,14 @@ class SubModel {
    */
   void addMessage(const std::string& message);
 
+  /**
+   * @brief getter to tell whether or not simulation generates a timeline
+   * @return true when simulation generates a file resuming the timeline; false when timeline has not been asked from jobs file
+   */
+  inline bool hasTimeline() const {
+    return timeline_ ? true : false;
```
Comment (lecourtoisflo) [inline dynawo/sources/Modeler/Common/DYNSubModel.h:738]: There is an API for shared_ptr which directly returns this status (use_count == 0)

**Review by lecourtoisflo — CHANGES_REQUESTED** [review]

**Comment by rosiereflo** [comment]
Before I review, please update the method used for Modelica models

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 2 above)
Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:738]: tell me more... You mean I may write "return timeline_.use_count() ? true : false ; }".  The compiler will do the job as I wrote it, no (?)

Comment (lecourtoisflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:738]: you can write directly "return timeline_.use_count() > 0;" or even "return timeline_" It is for readability. There is never need to use "(sth...) ? true : false" because it means that you can just use "(sth)"

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 1 above)
Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: Yes, but it depends upon who reads the line. As rules are not strictly defined, I won't discuss the coding style. I prefer to ask clang-format to do the job in an objective way than discussing about how to write code. In that particular case it was easier to write since that condition appears in other places

Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: anyway I would prefer a solution in which this test is included in the addEvent method.

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 2 above)
Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:738]: You are right, I did not think of it :( let's keep it this way. I have enough trouble pushing changes... I'm not sure "return timeline_;" works since timeline_ is not of bool type, it's a class. Compiler may be smart enough to turn it implicitly to a bool since it is a bool function; anyhow it won't run faster if I change if, for sure.

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 1 above)
Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: Florentine, it would have been more clear and a lot simpler. I would have done it but you asked not to call addEvent() when it was not usefull. In this way we don't call addEvent() when it is not needed, saving some little time, but code is more complex. In the addEvent() I add an assert() that tells not to call addEvent().

Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: I think there is a misunderstanding: I asked to not build the TimelineMessage when not needed, not to avoid calling addEvent. You can also change the addEvent parameters. Please update this and also the associated Modelica model equivalent methods as today this development cannot working.

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 2 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:738]: I agree with Florent, so as anyway there will be an update let's update this.

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 1 above)
Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: Also, please amend the commit so that its description explains what it does

Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: commit is amended in github

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 2 above)
Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:738]: I can change "return timeline_ ? true : false;" to "return timeline_.use_count() >0;" . In the first cas the compiler and boost chooses the best way to treat the line. In the second Florent tells the compiler what to do. My opnion is to always let compiler do optimization. The third way "return timeline_;" does something else.

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 1 above)
Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: Also, I don't get you. addEvent() is the way to build a Timeline message. I avoid calling it when not needed. Sometimes addEvent() has a parameter of type MessageLine that is build inside. avoiding addEvent() will avoid calling MessageLine(). In the case I let MessageLine() build a MessageLine, it is something I did not see. I make a review about it.

Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: Let me rephrase: I would like a solution in which the hasTimeline test is inside the addEvent method and we do not build the instance of TimelineMessage if !hasTimeline

Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: regarding the commit message, adding a message on github is not equivalent. Please use the command 
git commit --amend
and then it will be updated in this PR at the next push you will make

Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: OK; regarding what we said, I re-wrote a lot, almost all the code I pushed. It will be easier to delete the branch s I can start a brand new one. that on was WIP and not supposed to be merged; a new branch will please everybody. That is why this push is on a date, like a draft, not supposed to be kept. Now it is quite hard for me to add something in the branch; I tried this morning and received an all bunch of injuries from git :(

Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: git push --force-with-lease

**Comment by e2l-su** [comment]
It looks like coverage fails since it is not available with --std=c++2a. Who asked to compile in C++20 ? not me, it should only compile in c++98

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp [hunk 3]
```diff
@@ -1186,7 +1186,9 @@ SubModel::addMessage(const string& message) {
 
 void
 SubModel::addEvent(const string& modelName, const MessageTimeline& messageTimeline) {
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNSubModel.cpp:1188]: We still build the MessageTimeline instance here. 
What I was proposing to try is to create 2 versions of addEvent:
SubModel::addEvent(const string& modelName, const string& message)
SubModel::addEvent(const string& modelName, KeyTimeline_t key, ...)
the add Event body will then be something like:
if (hasTimeline()) {
messageTimeline messageTimeline(...)
timeline_->addEvent(getCurrentTime(), modelName, messageTimeline.str(), messageTimeline.priority())
}
and then to call those in the models instead of the current version of addEvent

File: dynawo/sources/Modeler/Common/DYNSubModel.h [hunk 4]
```diff
@@ -730,6 +730,14 @@ class SubModel {
    */
   void addMessage(const std::string& message);
 
+  /**
+   * @brief getter to tell whether or not simulation generates a timeline
+   * @return true when simulation generates a file resuming the timeline; false when timeline has not been asked from jobs file
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNSubModel.h:735]: "false when timeline has not been asked from jobs file" => "false otherwise"

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp [hunk 5]
```diff
@@ -693,14 +693,14 @@ ModelBus::evalZ(const double& /*t*/) {
     topologyModified_ = true;
     if (currState == OPEN) {
       switchOff();
-      network_->addEvent(id_, DYNTimeline(NodeOff));
+      if (network_->hasTimeline()) network_->addEvent(id_, DYNTimeline(NodeOff));
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp:696]: useless tests

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 6]
```diff
@@ -702,7 +696,7 @@ Simulation::init() {
   Trace::info() << DYNLog(CurveInit) << Trace::endline;
   Trace::info() << "-----------------------------------------------------------------------" << Trace::endline;
   const std::vector<double>& y = solver_->getCurrentY();
-  unsigned nbCurves = 0;
+  unsigned int nbCurves = 0;
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:699]: useless

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 7]
```diff
@@ -994,7 +989,7 @@ Simulation::terminate() {
     fileCurves.close();
   }
 
-  if (timelineOutputFile_ != "") {
+  if (hasTimeline() && timelineOutputFile_ != "") {
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:992]: hasTimeline() test should be enough

File: dynawo/sources/Simulation/DYNSimulation.h [hunk 8]
```diff
@@ -212,13 +213,23 @@ class Simulation {
...
    */
   inline void setTimelineExportMode(const exportTimelineMode_t& mode) {
     exportTimelineMode_ = mode;
   }
+
+  /**
+   * @brief getter for the export mode of timeline
+   * @return timeline's export mode
+   */
+  inline const exportTimelineMode_t& getTimelineExportMode() const {
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.h:229]: no need of const & for an enum (same above)

File: dynawo/sources/Simulation/DYNSimulation.h [hunk 9]
```diff
@@ -585,7 +598,16 @@ class Simulation {
    * @brief configure the final state outputs
    */
   void configureFinalStateOutputs();
-};
+
+  /**
+   * @brief getter which tells whether or not simulation should generate a timeline
+   * @return true when simulation should generate a file resuming the timeline; false when timeline has not been asked
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.h:604]: false otherwise

File: dynawo/sources/Solvers/Common/DYNSolverImpl.h [hunk 10]
```diff
@@ -288,6 +288,15 @@ class Solver::Impl : public Solver, private boost::noncopyable {
    */
   void printUnstableRoot(double t, const std::vector<state_g>& G0, const std::vector<state_g>& G1) const;
 
+  /**
+   * @brief getter to tell whether or not simulation generates a timeline
+   * @return true when simulation generates a file resuming the timeline; false when timeline has not been asked from jobs file
```
Comment (rosiereflo) [inline dynawo/sources/Solvers/Common/DYNSolverImpl.h:293]: same as above

**Review by rosiereflo — CHANGES_REQUESTED** [review]

**Comment by rosiereflo** [comment]
MR has been cleaned up, please now follow the usual process of dynawo: 1) commit, 2) push, 3)review, 4) modif on same branch, 5) commit on same branch, 6) push and go back to step 3

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 4 above)
Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.h:735]: ok

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp (hunk 5 above)
Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp:696]: That is to prevent building a DYNTimeline(...) object. It is exactly the same as
if ( network_->hasTimeline()) {
      DYNTimeline_object TL( ....);
      network_->addEvent( id_, TL) ;
}
It is exactly what is asked in the preceding lines with MessageTimeline. I don't know what to do

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 6 above)
Comment (e2l-su) [reply dynawo/sources/Simulation/DYNSimulation.cpp:699]: ok

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp (hunk 5 above)
Comment (rosiereflo) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp:696]: please see my proposal above

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 7 above)
Comment (e2l-su) [reply dynawo/sources/Simulation/DYNSimulation.cpp:992]: ok

File: dynawo/sources/Simulation/DYNSimulation.h (hunk 8 above)
Comment (e2l-su) [reply dynawo/sources/Simulation/DYNSimulation.h:229]: of course... But looking at an object exportTimelineMode_t one does not know if it is an enum or not. If it turns to be something else later ( an enum class or even a const object ) that code will fit. I don't think I used that getter anyway. It is because it should always be like that...

File: dynawo/sources/Solvers/Common/DYNSolverImpl.h (hunk 10 above)
Comment (e2l-su) [reply dynawo/sources/Solvers/Common/DYNSolverImpl.h:293]: done

File: dynawo/sources/Simulation/DYNSimulation.h (hunk 8 above)
Comment (rosiereflo) [reply dynawo/sources/Simulation/DYNSimulation.h:229]: If not used, please remove the method, otherwise please remove the reference as it should not be used with fundamental data types (and if we change it to a class then this should be updated)

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp (hunk 1 above)
Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelLine.cpp:981]: it works now

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp (hunk 3 above)
Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:1188]: I think I get your point... Tell me if what you ask is to get rid of the class called MessageTimeline ? Or maybe you don't want that a Timeline object/reference/pointer is passed as an argument to addEvent() (?) Until now I understood that you  did not want to create a MessageTimeline before/when calling addEvent. Not calling addEvent resolved that last point.

File: dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp (hunk 5 above)
Comment (e2l-su) [reply dynawo/sources/Models/CPP/ModelNetwork/DYNModelBus.cpp:696]: ok, I close resolve this discussion since it points to anotherone

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp (hunk 3 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:1188]: I would like to try a solution as follows:

On model side
network-> addEvent(... ) (no test of hasTimeline, way better for readability and maintenability)

On add Event side :
if (hasTimeline()) {
messageTimeline messageTimeline(...)
timeline_->addEvent(getCurrentTime(), modelName, messageTimeline.str(), messageTimeline.priority())
}

Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:1188]: On Model side, addEvent will be addEvent( string, string) or addEvent( string, key_t) ...
And I will pass the arguments to the ctor of messasgeTimeline only in cas we need to log an event. That's really easy to do, why didn't you tell ?

Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:1188]: no comment

Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp]: done, I even added a unitary test
